jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -292,23 +308,31 @@ private void pullBaseImage() {
   private void obtainBaseImageLayers(boolean layersRequiredLocally) {
     ProgressEventDispatcher.Factory childProgressDispatcherFactory =
         Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
-
     results.baseImageLayers =
         executorService.submit(
-            () ->
-                scheduleCallables(
-                    layersRequiredLocally
-                        ? ObtainBaseImageLayerStep.makeListForForcedDownload(
-                            buildContext,
-                            childProgressDispatcherFactory,
-                            results.baseImageAndRegistryClient.get().images.get(0),
-                            results.baseImageAndRegistryClient.get().registryClient)
-                        : ObtainBaseImageLayerStep.makeListForSelectiveDownload(
-                            buildContext,
-                            childProgressDispatcherFactory,
-                            results.baseImageAndRegistryClient.get().images.get(0),
-                            results.baseImageAndRegistryClient.get().registryClient,
-                            results.targetRegistryClient.get())));
+            () -> {
+              Map<Image, List<Future<PreparedLayer>>> baseImageLayers = new HashMap<>();
+              for (Image image : results.baseImageAndRegistryClient.get().images) {
```

Comment:
I realized we could have increased concurrency further. `PullBaseImageStep` is now designed to return all manifests all at once, meaning that downloading base image layers can start only after _all_ manifests have been downloaded. Moreover, in `PullBaseImageStep`, we will probably end up having to download all manifests synchronously and sequentially. Theoretically and ideally we should be able to concurrently download manifests as well as start downloading base image layers while there are still pending manifest downloads. However, because this requires significant redesign of `PullBaseImageStep` (i.e., we should schedule multiple threads (multiple step instances) each downloading an assigned manifest, resulting in `List<Future<Image>>` instead of the current `List<Image>`) and because manifests are small and generally quick to download, I think the current implementation is reasonable. We just need to keep track of this issue so that we can eventually implement this.

Reply:
This is a very good point .I hadn't realized that we are using a future<List<Image>> instead of List<Future<Image>> . As you said the manifest files are really small and thus there is a lower chance of them slowing down the build process I support the fact that we can refactor this later.I have put this point on the list for items to be improved once a working multi-image is completed.

Reply:
> because manifests are small and generally quick to download, I think the current implementation is reasonable.

For your insight, this is somewhat acceptable, but we've seen many, many Jib users working in a slow network where it takes, e.g., 4~5 seconds to download a manifest. So, if you are sequentially downloading 3 manifests, they will add up to ~15 seconds. 15 seconds would still be a small fraction considering that the entire build would take far more time in such a slow network, but it will be a significant bottleneck, e.g., when not pushing to a remote registry.

Reply:
Leaving this comment thread open, for easy reference in the future. You don't have to obsessed with collapsing every comment thread.

Reply:
I always thought it was a requirement to close all of them hahaha I will leave this one open